### ISSUE TITLE
[Suggestion] Blinker/repeater/clock circuit

### ISSUE THREAD

**Issue by Calvert4096** [tsunamayo/Starship-EVO#2509]
### Short Description of the issue:

I believe in this thread https://github.com/tsunamayo/Starship-EVO/issues/1264 talrey mentioned a "repeater" circuit is possible using current logic.  There's some discussion on Discord to this end, but the only way I've been able to get this to work is with a switch -> and/or gates-> rotor -> query gate -> output (e.g. light) -> back to and/or gates, as shown:

![image](https://user-images.githubusercontent.com/65469633/85367378-f6d2be80-b4dd-11ea-9ca1-130b37919f15.png)

It seems like this *should* be achievable with something like this setup (no rotors, and no *clunk* *clunk* clunk*!).  However this circuit does not repeat after the first passthrough.  I'm curious if the repeating behavior is arrested intentionally as a way to prevent players from making laggy builds.

![image](https://user-images.githubusercontent.com/65469633/85367472-284b8a00-b4de-11ea-9c08-222273f2daf4.png)

**Comment by tsunamayo** [comment]
Yes I think that was the fix I did to address some logic time bomb people dropped at me.
I could either make you a repeater or do a complete overhaul of the logic engine, I think I will work on a repeater instead ;)
I keep you posted!

**Comment by milan-t** [comment]
It is possible create some astable multivibrator with logic gates.
Simple inverting logic gate + delay gate as feedback is enough to create this kind of oscilator.
With other logic gate it can be turned on and off.

Examples:
**1. AND + NOR (working as logical Not) + Delay**
![simple_astable_circuit](https://user-images.githubusercontent.com/65783423/85445340-91410980-b593-11ea-970f-b26ec3fb4a9a.png)
_(Top line pictures shows connections, bottom line pictures blinking when the switch is on.)_

![circuit_explanation](https://user-images.githubusercontent.com/65783423/85449075-a15ae800-b597-11ea-87be-bf3299ff490a.png)

**2. Blinker circuit in my hoover bus**
NAND + Delay + OR (as on/off) + other 2 ANDs controls "blinkers"

Actual look:
![blinker_circuit_hoover_bus](https://user-images.githubusercontent.com/65783423/85449651-483f8400-b598-11ea-8aaf-f33ea8eba727.png)
_(blinker circuit is composed of orange gates + grass green OR gate)_

Schematic:
![Blinker_circuit_explanation_](https://user-images.githubusercontent.com/65783423/85450601-39a59c80-b599-11ea-8bc6-9c07a086760b.png)

**Comment by milan-t** [comment]
Another idea. If is needed some different duty cycle then 50 % (1:1).
Combine NAND+delay "clock" circuit with an other delay gate through AND gate to create for example strobe light:
![strobe_light_circuit](https://user-images.githubusercontent.com/65783423/85455477-37920c80-b59e-11ea-9394-f3b56c32ff84.png)

**Comment by milan-t** [comment]
@Calvert4096 I examined your switch based repeater and I think that problem is how switches works and OR gates are used. The switch flips when signal changes from logical 0 to 1 (on rising edge). Reason why first time it does something is probably how gates works and propage signal. Probably all logic gates have output set to 0 before they receive any input change from user (via button, switch, slider or query gate).

If I'm right, then the circuit behavior can be described in the following way:
1. when button is pressed, its output changes from 0 to 1
2. first OR gate has now inputs 0+0+1 which means that OR gate output is 1
3. first switch flips
4. first NAND stays 0, delay gate delays change from 0 to 1 for 1 second
5. meanwhile button is released, so its output changes back to 0 
6. now first OR has all inputs 0 therefore output is also 0
7. after 1 second first delay gate has 1 as output, this means that second OR gate has value 1 (1+0)
8. change of second OR gate from 0 to 1 flips second switch
9. second NAND stays 0 
10. after 1 second output of second delay gate changes to 1
11. this change means that first OR gate now has inputs 1 + 0 + 0 (2nd delay+2nd NAND+button) so it is again change from 0 to 1 and first switch flips
12. unfortunately first NAND changes from 0 to 1 but 1st delay has still value 1, second OR stays 1 (1+1, later 0+1)
13. after 1 second delay changes to 0 but at least one input of second OR gate is 1 therefore OR gate stays 1
And this is final state.

Here is timing diagram of what I tried to describe:
![switch_repeater_timing_diagram](https://user-images.githubusercontent.com/65783423/85471719-243c6c80-b5b1-11ea-808c-6db7a63d7861.png)

It stucks because there is no change from 0 to 1 on OR 1 or OR 2 later that can flip first or second switch.

Sorry for exhausting description.

**Comment by tsunamayo** [comment]
@milan-t @Calvert4096 if that helps I can revert back the switch behavior to full propagation, (so both 0 -> 1 and 1 -> 0)... I think it is a bit confusing.

**Comment by milan-t** [comment]
For me current behavior of the switch works great. It makes possible to control mechanisms like doors by buttons.

Because the switch works like T flip-flop (term looked up in Wikipedia) and can be used as clock frequency divider (by 2) that opens possibility to create binary counters.

If the switch flips on both rising and falling edge of signal (reacts to both input state changes), it will copy or invert what it has on input. Then no use for button toggled mechanisms like doors, if I'm not mistaken.

**Comment by tsunamayo** [comment]
yes, that the debate, current behaviour opens possibility for power user, but is confusing for the 95% other common folks...

**Comment by Calvert4096** [comment]
If that's the case, I like the idea of leaving the behavior in such a way that rewards "power users."  If someone is sufficiently curious, I'm sure it will eventually be documented in the wiki or on youtube and they can learn as much with a quick search.

**Comment by Calvert4096** [comment]
@milan-t  I attempted to recreate two of your blinker circuits and I'm unable to get it to oscillate.  In both cases the light just stays on.

@tsunamayo Does milan-t's clock circuit work for you?  I'm wondering if this is in fact a bug.

**Comment by milan-t** [comment]
@Calvert4096 It is strange that is not working.

To make sure that pulse generators (oscillators) works I started new game and created them again.

Here is a screenshot with annotations:
![pulse_generator_examples_with_connections](https://user-images.githubusercontent.com/65783423/85918965-9b028f80-b867-11ea-946a-cb54c1cf44c9.png)

Attaching also a blueprint:
[pulse_generator_examples_blueprint.zip](https://github.com/tsunamayo/Starship-EVO/files/4840228/pulse_generator_examples_blueprint.zip)

I played little more with oscillators and I found out that it is necessary to make all connection right before activating circuit with a switch. I ended up with not working oscillator when I did not link NAND gate back to delay and turned swicth on and off. Later when I added missing link, the circuit was still broken.

It is possible to recover from this state by changing NAND gate to other gate type and back.
Other way how to fix problem is to remove link from delay to NAND, turn the switch on an off, link delay to NAND again.
Last thing I tried is simply remove delay and NAND gate. Then add them back and wire them this time correctly before activation.

I hope that helps a little.

**Comment by Calvert4096** [comment]
@milan-t Thanks for the detailed explanation!  I built your configurable pulse generator and it works nicely.  I'm not sure why my previous attempts at reproduction didn't work, but, as you say, it may well have to do with attempting test pieces of the circuit piecemeal before the whole thing is wired up.

@tsunamayo If gates have some hidden internal state that would affect whether a pulse  generator would work, maybe it's worth making that state apparent to the player (some kind of built-in indicator?).  It seems like there's ample opportunity for a player to attempt to re-wire existing circuits, and find it won't work even if they follow a guide like the one milan-t provided, causing frustration.  They would belatedly discover they have to rebuild a circuit from a blank slate, because there's no way to tell which gates' hidden states are preventing (for example) a pulse generator from working.  Making the circuit elements' internal state transparent to the player may help to prevent this.

**Comment by MatCat** [comment]
If you want a clock that can be turned on and off with a switch I came up with the simple circuit of hooking up a switch to a simple pulse generator circuit, the pulse generator is set to 0.1, and the clock is only stable with 0.2 and higher.  This works well in that from a fresh built state this will allow a clock circuit to be started or stopped with just a simple switch.

![Screenshot 2020-09-08 21 01 23](https://user-images.githubusercontent.com/598987/92552891-899a1400-f216-11ea-9c6a-253660536539.png)

**Comment by MatCat** [comment]
Doh for some reason its not showing in the screenshot arrow from right clock down to the AND gate :)
